batch/jvm-entryway/src/main/scala/is/hail/QoBAppender.scala

```diff
@@ -0,0 +1,83 @@
+// Do not move this to a different package without updating packages= in log4j2.properties
+package is.hail
+
+import java.io._
+import java.util.concurrent.TimeUnit
+import org.apache.logging.log4j.core._
+import org.apache.logging.log4j.core.layout._
+import org.apache.logging.log4j.core.appender._
+import org.apache.logging.log4j.core.config._
+import org.apache.logging.log4j.core.config.plugins._
+import scala.collection.mutable
+
+object QoBOutputStreamManager {
+  private var _instances: mutable.Map[Layout[_], QoBOutputStreamManager] = mutable.Map()
+
+  def getInstance(layout: Layout[_]): QoBOutputStreamManager = synchronized {
+    _instances.getOrElseUpdate(layout, new QoBOutputStreamManager(layout))
```

Comment:
Does this create a new object only to likely destroy it immediately after? Is that something to worry about? Unclear to me whether any of these log4j classes register themselves somewhere upon initialization.

Reply:
`map.getOrElseUpdate` is like Java's `computeIfAbsent`. If the key exists it returns the value. Otherwise, it evaluates the thunk argument, inserts it into the map, and returns it.

`getInstance` is called in the constructor to `QoBAppender`. `QoBAppender` is created by the Log4J2 infrastructure via `createAppender`. Log4J2 finds that method by way of all these annotations (`@PluginFactory`, etc.).